### PR TITLE
Item-Collection mapper

### PR DESCRIPTION
**Pull request by Atmire-Kristof**
This feature is based on the “Mapped collections” feature in DSpace6.

REST API Implementation: https://github.com/DSpace/DSpace/pull/2282

**collection-item mapper**
The route to this page can be found at `/collections/<uuid>/edit/mapper`. Here, the user (when logged in), has the ability to map items to the current collection.
The page exists out of a description and two tabs labelled “Browse” and “Map”.
The “Browse” tab contains items already mapped to the collection and the option to un-map items by selecting them and pressing the confirm button. The “Map” tab contains all items not yet mapped to or part of the collection and the option to map items by selecting them and pressing the confirm button.
You’ll also notice the “Map” tab has a search feature, as this page will usually contain a lot of items.

**item-collection mapper**
The route to this page can be found at `/items/<uuid>/edit/mapper` or by navigating to the edit page and pressing the “Mapped collections” button.
Just like the collection-item mapper, this page contains two tabs, “Browse” and “Map”.
The “Browse” tab contains all collections the item is mapped to. The “Map” tab contains all collections the item is not yet mapped to or part of.
Both also have the option to select collections for un-mapping or mapping the item.

**object-select**
An abstract object-select component was added, which can be extended depending on the type of object the list is composed of. For this PR, two components exist that extend this one: collection-select and item-select.

_input_
- key: This should be a unique key used for storing the selected objects into the store. This way they don’t interfere with one another.
- dsoRD$: Observable<RemoteData<PaginatedList>> with TDomain being any type of object defined in the child component
- paginationOptions: The PaginationComponentOptions used for the list
- confirmButton: A message key used to display on the confirmation button
- hideCollection: a boolean used to disable the collection column

_output_
- confirm: A list of UUIDs of the selected objects after pressing the confirmation button (string[])

### CODE REVIEW

File: config/environment.default.js [hunk 1]
```diff
@@ -18,9 +18,17 @@ module.exports = {
   // Caching settings
   cache: {
     // NOTE: how long should objects be cached for by default
-    msToLive: 15 * 60 * 1000, // 15 minutes
+    msToLive: {
+      default: 15 * 60 * 1000, // 15 minutes
+      exportToZip: 5 * 1000 // 5 seconds
```
Comment (tdonohue) [inline config/environment.default.js]: It's unclear to me what these new settings in `environment.default.js` are for. Are these an unrelated change, or maybe local configuration?

File: src/app/+collection-page/collection-item-mapper/collection-item-mapper.component.ts [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+  styleUrls: ['./collection-item-mapper.component.scss'],
+  templateUrl: './collection-item-mapper.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  animations: [
+    fadeIn,
+    fadeInOut
+  ]
+})
+/**
+ * Collection used to map items to a collection
```
Comment (tdonohue) [inline src/app/+collection-page/collection-item-mapper/collection-item-mapper.component.ts]: This comment doesn't make sense. Should this say "Component used to map items to a collection"?

File: src/app/+search-page/search-filters/search-filter/search-facet-filter/search-facet-filter.component.ts [hunk 3]
```diff
@@ -126,7 +126,7 @@ export class SearchFacetFilterComponent implements OnInit, OnDestroy {
         this.animationState = 'ready';
         this.filterValues$.next(rd);
       }));
-      this.subs.push(newValues$.pipe(first()).subscribe((rd) => {
+      this.subs.push(newValues$.pipe(take(1)).subscribe((rd) => {
```
Comment (tdonohue) [inline src/app/+search-page/search-filters/search-filter/search-facet-filter/search-facet-filter.component.ts]: I'm not sure I understand the reason why we are switching from `first()` to `take(1)`?  If this is necessary, we should also remove `first` from the `import` statement above

File: src/app/+search-page/search-filters/search-filter/search-filter.component.ts [hunk 4]
```diff
@@ -37,7 +37,7 @@ export class SearchFilterComponent implements OnInit {
    * Else, the filter should initially be collapsed
    */
   ngOnInit() {
-    this.getSelectedValues().pipe(first()).subscribe((isActive) => {
+    this.getSelectedValues().pipe(take(1)).subscribe((isActive) => {
```
Comment (tdonohue) [inline src/app/+search-page/search-filters/search-filter/search-filter.component.ts]: Another switch from `first` to `take(1)`.  Again, if this is needed, remove `first` from the import

File: src/app/+search-page/search-filters/search-filter/search-facet-filter/search-facet-filter.component.ts (hunk 3 above)
Comment (artlowel) [reply src/app/+search-page/search-filters/search-filter/search-facet-filter/search-facet-filter.component.ts]: `first()` and `take(1)` are nearly identical, the key difference is that `first()` will expect to get at least one value, and will throw an EmptyError if the observable completes without firing one. `take(1)` will simply never execute the subscribe if there are no values. @Atmire-Kristof was getting these EmptyErrors in certain cases, and that's why we replaced them.

We'll remove the import for first

File: src/app/+search-page/search-filters/search-filter/search-filter.component.ts (hunk 4 above)
Comment (artlowel) [reply src/app/+search-page/search-filters/search-filter/search-filter.component.ts]: Will do

File: src/app/+collection-page/collection-item-mapper/collection-item-mapper.component.ts (hunk 2 above)
Comment (artlowel) [reply src/app/+collection-page/collection-item-mapper/collection-item-mapper.component.ts]: It should. We'll change it

File: config/environment.default.js (hunk 1 above)
Comment (artlowel) [reply config/environment.default.js]: This PR includes the same response cache refactor @LotteHofstede's #349 PR is based on. It adds the ability to cache certain responses for different lengths of time. (e.g. you may want to cache most responses for 15 minutes, but for certain admin features you may want to disable the cache, or cache it a very short time) However the exportToZip config _is_ something that snuck in from an unrelated branch i.e. the export to zip feature that is on hold atm waiting for the [scripts and processes work](https://github.com/DSpace/Rest7Contract/pull/17) on the REST side. We'll remove it

**Comment by tdonohue** [comment]
I tried to review this PR again today, but I'm having a hard time doing a code review. It looks like it contains most of the changes in #349, but seems to have possibly diverged at some point... so attempting to diff this branch with that of #349 doesn't seem to be working.

I suspect we need to get #349 updated/merged quickly (it's nearly there) and then rebase this PR to make it easier to review.

**Comment by Atmire-Kristof** [comment]
The latest changes to this PR are dependant on the following REST API PR:
https://github.com/DSpace/DSpace/pull/2282

File: src/app/core/data/mapping-collections-reponse-parsing.service.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import { Injectable } from '@angular/core';
+import { ResponseParsingService } from './parsing.service';
+import { RestRequest } from './request.models';
+import { DSpaceRESTV2Response } from '../dspace-rest-v2/dspace-rest-v2-response.model';
+import { PaginatedList } from './paginated-list';
+import { PageInfo } from '../shared/page-info.model';
+import { ErrorResponse, GenericSuccessResponse, RestResponse } from '../cache/response.models';
+
+@Injectable()
+export class MappingCollectionsReponseParsingService implements ResponseParsingService {
```
Comment (tdonohue) [inline src/app/core/data/mapping-collections-reponse-parsing.service.ts]: Please add TypeDocs to this new Service. This should likely also be renamed "MappedCollectionsResponseParsingRequest" (the REST endpoint is now named /mappedCollections)

File: src/app/core/data/request.models.ts [hunk 6]
```diff
@@ -185,6 +186,12 @@ export class BrowseItemsRequest extends GetRequest {
   }
 }
 
+export class MappingCollectionsRequest extends GetRequest {
```
Comment (tdonohue) [inline src/app/core/data/request.models.ts]: Also needs Typedocs.  This should likely also be renamed "MappedCollectionsRequest" (the REST endpoint is now named `/mappedCollections`)

File: src/app/core/data/collection-data.service.ts [hunk 7]
```diff
@@ -57,4 +64,55 @@ export class CollectionDataService extends ComColDataService<Collection> {
...
   }
+
+  /**
+   * Fetches the endpoint used for mapping items to a collection
+   * @param collectionId   The id of the collection to map items to
+   */
+  getMappingItemsEndpoint(collectionId): Observable<string> {
+    return this.halService.getEndpoint(this.linkPath).pipe(
+      map((endpoint: string) => this.getIDHref(endpoint, collectionId)),
+      map((endpoint: string) => `${endpoint}/mappingItems`)
```
Comment (tdonohue) [inline src/app/core/data/collection-data.service.ts]: I believe this endpoint path is incorrect. It's `/mappedItems`.  See https://github.com/DSpace/DSpace/pull/2282

Comment (Atmire-Kristof) [reply src/app/core/data/collection-data.service.ts]: > I believe this endpoint path is incorrect. It's `/mappedItems`. See [DSpace/DSpace#2282](https://github.com/DSpace/DSpace/pull/2282)

Updated the endpoint's path, as well as all methods related to the endpoint, for more consistency.
While testing, I also discovered a bug in the collection-list on the item-collection-mapper component. Included a fix for that.

**Comment by Atmire-Kristof** [comment]
> @Atmire-Kristof : I've tested this again today, and run into a few small bugs:
> 
> 1. The `/collections/[uuid]/mapper` path doesn't appear to work. I'm getting a large number of errors in the console that say "Can't set headers after they are sent", and the browser just shows a blank page.
> 2. The '/items/[uuid]/edit/mapper` path works overall. I can map an Item to a Collection.  However _all_ Collections are listed, including the Item's current Collection.  If I select the current Collection and attempt a mapping it fails (obviously).  But, mapping to other collections work -- however the usability is a bit odd.  After clicking the button "Map item to selected collections", I see a green notice, but I still remain on the "Map" tab and the mapped collections disappear.  I'd expect to be sent to the "Browse" tab to show the list of collections it's now successfully mapped to.

1. We weren't able to reproduce this issue.
2. The code was excluding already mapped collections from the "Map" tab, but I forgot to also exclude the owning collection of the item. This should now be fixed. After a mapping request has finished, the view now switches to the first tab (on both item-collection-mapper and collection-item-mapper).

**Comment by Atmire-Kristof** [comment]
The currently failing Travis build seems to be an issue that has to be fixed on their side.
Similar issue (with the exact same error message) from a while ago: https://travis-ci.community/t/packages-cannot-be-authenticated-gpg-error-google-chrome-stable/3000

File: src/app/+search-page/search-service/search.service.ts [hunk 8]
```diff
@@ -373,6 +373,15 @@ export class SearchService implements OnDestroy {
     return '/search';
   }
 
+  /**
+   * Clear all request cache related to discovery objects
+   */
+  clearDiscoveryRequests() {
```
Comment (artlowel) [inline src/app/+search-page/search-service/search.service.ts]: I don't like this: you should never have to manually clear the cache. Instead, use requests with a responseMsToLive that is lower or 0. Then, in requestService, if a request for a URL is cached, a new request is made with a lower responseMsToLive value than the cached version, update that value in the cache before you check the validity. That should bypass the need for this method
